File: server/cmd/seed/seed.go
package main

import (
	"context"
	dbsql "database/sql"
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"

	"github.com/emonadeo/gdol/pkg/generated/sqlc"

	_ "github.com/lib/pq"
)

type User struct {
	Banned bool
	Id     int
	Name   string
}

type Level struct {
	Id          int      `json:"id"`
	Level_id    int      `json:"level_id"`
	Name        string   `json:"name"`
	Position    int      `json:"position"`
	Requirement int      `json:"requirement"`
	Publisher   User     `json:"publisher"`
	Creators    []User   `json:"creators"`
	Verifier    User     `json:"verifier"`
	Video       string   `json:"video"`
	Records     []Record `json:"records"`
}

type Record struct {
	Id       int    `json:"id"`
	Player   User   `json:"player"`
	Progress int    `json:"progress"`
	Status   string `json:"status"`
	Video    string `json:"video"`
}

func main() {
	ctx := context.Background()

	db, err := dbsql.Open("postgres", "user=example password=example dbname=gdol sslmode=disable")
	if err != nil {
		panic(err)
	}

	queries := sqlc.New(db)

	tx, err := db.Begin()
	if err != nil {
		panic(err)
	}

	file, err := filepath.Abs("pkg/generated/pointercrate/_list.json")
	if err != nil {
		panic(err)
	}
	listRaw, err := os.ReadFile(file)
	if err != nil {
		panic(err)
	}
	list := []int{}
	err2 := json.Unmarshal(listRaw, &list)
	if err2 != nil {
		panic(err2)
	}

	currentLevelIds := []int64{}
	for i, v := range list {
		levelRaw, err := os.ReadFile("pkg/generated/pointercrate/" + strconv.Itoa(v) + ".json")
		if err != nil {
			panic(err)
		}
		level := Level{}
		err1 := json.Unmarshal(levelRaw, &level)
		if err1 != nil {
			panic(err1)
		}

		// Insert users
		publisherId := seedUser(queries, ctx, level.Publisher)
		verifierId := seedUser(queries, ctx, level.Verifier)
		creatorIds := seedUsers(queries, ctx, level.Creators)

		// Insert level
		levelId, err := queries.InsertLevel(ctx, sqlc.InsertLevelParams{
			Name:       level.Name,
			GdID:       int64(level.Level_id),
			UserID:     publisherId,
			VerifierID: verifierId,
			Video: dbsql.NullString{
				String: level.Video,
				Valid:  level.Video != "",
			},
			Requirement: dbsql.NullInt16{
				Int16: int16(level.Requirement),
				Valid: true,
			},
		})

		if err != nil {
			panic(err)
		}

		currentLevelIds = append(currentLevelIds, levelId)

		// Link creators
		for _, creatorId := range creatorIds {
			err := queries.InsertUserCreatedLevel(ctx, sqlc.InsertUserCreatedLevelParams{
				UserID:  creatorId,
				LevelID: levelId,
			})
			if err != nil {
				panic(err)
			}
		}

		// Insert level log
		_, err = queries.InsertListLog(ctx, sqlc.InsertListLogParams{
			Action:       sqlc.ListLogActionAdd,
			LevelID:      levelId,
			ListLevelIds: currentLevelIds,
			From: dbsql.NullInt16{
				Int16: 0,
				Valid: false,
			},
			To: dbsql.NullInt16{
				Int16: int16(i + 1),
				Valid: true,
			},
		})

		if err != nil {
			panic(err)
		}

		// Seed Records
		seedRecords(queries, ctx, levelId, level.Records)
	}

	tx.Commit()
}

func seedUser(qry *sqlc.Queries, ctx context.Context, user User) int64 {
	userExists, err := qry.HasUserWithName(ctx, user.Name)
	if err != nil {
		panic(err)
	}

	if userExists {
		id, err := qry.GetUserIdByName(ctx, user.Name)
		if err != nil {
			panic(err)
		}
		return id
	}

	id, err := qry.InsertUser(ctx, sqlc.InsertUserParams{
		Name:        user.Name,
		Nationality: dbsql.NullString{Valid: false},
	})
	if err != nil {
		panic(err)
	}
	return id
}

func seedUsers(qry *sqlc.Queries, ctx context.Context, users []User) []int64 {
	ids := []int64{}
	for _, user := range users {
		ids = append(ids, seedUser(qry, ctx, user))
	}
	return ids
}

func seedRecord(qry *sqlc.Queries, ctx context.Context, levelId int64, record Record) {
	recordExists, err := qry.HasRecord(ctx, sqlc.HasRecordParams{
		UserID:  int64(record.Player.Id),
		LevelID: levelId,
	})
	if err != nil {
		panic(err)
	}

	if recordExists {
		return
	}

	playerId := seedUser(qry, ctx, record.Player)

	err = qry.InsertRecords(ctx, sqlc.InsertRecordsParams{
		Percentage: dbsql.NullInt16{
			Int16: int16(record.Progress),
			Valid: true,
		},
		Video: dbsql.NullString{
			String: record.Video,
			Valid:  record.Video != "",
		},
		UserID:  playerId,
		LevelID: levelId,
	})
	if err != nil {
		panic(err)
	}
}

func seedRecords(qry *sqlc.Queries, ctx context.Context, levelId int64, records []Record) {
	for _, record := range records {
		seedRecord(qry, ctx, levelId, record)
	}
}

File: _seed.ts
import list from './_seed_data/_list.json' assert { type: 'json' };
import { Database } from './deps.ts';

const _json = { assert: { type: 'json' } } as const;
const _seed_data_dir = '_seed_data';

type User = {
	banned: boolean;
	id: number;
	name: string;
};

type Nationality = {
	country_code: string;
	nation: string;
	subdivision?: string;
};

type Record = {
	id: number;
	progress: number;
	status: 'approved' | 'rejected' | 'submitted' | 'under consideration';
	video?: string;
	player: User;
	nationality?: Nationality;
};

type Level = {
	name: string;
	position?: number;
	id: number;
	requirement: number;
	video?: string;
	level_id?: number;
	verifier: User;
	creators: User[];
	publisher: User;
	records: Record[];
	thumbnail?: string;
};

export async function seed(db: Database) {
	const level_ids: number[] = [];
	for (const file_id of list) {
		const level = (await import(`./${_seed_data_dir}/${file_id}.json`, _json)).default as Level;
		const level_id = seed_level(db, level);
		level_ids.push(level_id);

		db.prepare(
			`INSERT INTO list_log ("list_level_ids", "timestamp", "action", "level_id", "to", "reason")
		 VALUES (:list_level_ids, unixepoch('now'), :action, :level_id, :to, :reason)`
		).run({
			list_level_ids: JSON.stringify(level_ids),
			action: 'ADD',
			level_id,
			to: level_ids.length,
			reason: 'Seeded',
		});
	}
}

function seed_level(db: Database, level: Level): number {
	const publisher_id = seed_user(db, level.publisher);
	const verifier_id = seed_user(db, level.verifier);
	const creator_ids = level.creators.map((creator) => seed_user(db, creator));

	const result = db
		.prepare(
			`INSERT INTO levels (name, gd_id, user_id, verifier_id, video, requirement)
				  VALUES (:name, :gd_id, :user_id, :verifier_id, :video, :requirement) RETURNING id`
		)
		.value<[number]>({
			name: level.name,
			gd_id: level.level_id,
			user_id: publisher_id,
			verifier_id,
			video: level.video,
			requirement: level.requirement,
		});

	if (result === undefined) {
		console.error('inserting level returned undefined id');
		return -1;
	}

	const [id] = result;

	// seed creators
	creator_ids.forEach((creator_id) => seed_user_created_level(db, creator_id, id));

	// seed records
	level.records.forEach((record) => seed_record(db, id, record));

	return id;
}

function seed_record(db: Database, level_id: number, record: Record): void {
	const user_id = seed_user(db, record.player);
	db.prepare(
		`INSERT INTO records (timestamp, percentage, video, user_id, level_id)
		 VALUES (unixepoch('now'), :percentage, :video, :user_id, :level_id)`
	).run({
		percentage: record.progress,
		video: record.video,
		user_id,
		level_id,
	});
}

function seed_user_created_level(db: Database, user_id: number, level_id: number): void {
	db.prepare('INSERT INTO user_created_level (user_id, level_id) VALUES (?, ?)').run(
		user_id,
		level_id
	);
}

function seed_user(db: Database, user: User): number {
	const find_result = db.prepare('SELECT id FROM users WHERE name = ?').value<[number]>(user.name);

	if (find_result !== undefined) {
		return find_result[0];
	}

	const insert_result = db
		.prepare(`INSERT INTO users (name) VALUES (?) RETURNING id`)
		.value<[number]>(user.name);

	if (insert_result === undefined) {
		console.error('inserting user returned undefined id');
		return -1;
	}

	const [id] = insert_result;
	return id;
}
